dags/sentinel2_gap_report.py

```diff
@@ -0,0 +1,87 @@
+"""
+# Generate a gap report between sentinel-cogs and deafica-sentinel-2 buckets
+
+This DAG runs once a month and creates a gap report in the folowing location:
+s3://deafrica-sentinel-2/monthly-status-report
+"""
+import json
+import csv
+from pathlib import Path
+from datetime import datetime
+
+from airflow import configuration
+from airflow import DAG
+from airflow.operators.python_operator import PythonOperator
+from utils.inventory import s3
+
+default_args = {
+    'owner': 'Airflow',
+    "start_date": datetime(2020, 7, 24),
+    'email': ['toktam.ebadi@ga.gov.au'],
+    'email_on_failure': True,
+    'email_on_success': True,
+    'email_on_retry': False,
+    'retries': 0,
+    'manifest_suffix': "manifest.json",
+    'africa_tiles': "data/africa-mgrs-tiles.csv",
+    'africa_conn_id': "deafrica-prod-migration",
+    "us_conn_id": "deafrica-migration_us",
+    "dest_bucket_name": "s3://deafrica-sentinel-2-inventory",
+    "src_bucket_name": "s3://sentinel-cogs-inventory",
+    "reporting_bucket": "s3://deafrica-sentinel-2",
+    "reporting_prefix": "monthly-status-report/",
+    "schedule_interval": "@monthly"
+}
+
+def generate_bucket_diffs():
+    """
+    Compare Sentinel-2 buckets in US and Africa and detect differences
+    A report containing missing keys will be written to s3://deafrica-sentinel-2/monthly-status-report
+    """
+    url_source = default_args['src_bucket_name']
+    url_destination = default_args['dest_bucket_name']
+    suffix = default_args['manifest_suffix']
+    africa_tile_ids_path = tile_ids_filepath = Path(configuration.get('core', 'dags_folder')). \
+                        parent.joinpath(default_args['africa_tiles'])
+    source_keys = []
+    destination_keys = []
+
+    # Read Africa tile ids
+    africa_tile_ids = []
+    with open(africa_tile_ids_path, 'r') as file:
+        ids = csv.reader(file)
+        africa_tile_ids = [row[0] for row in ids]
+
+    s3_inventory = s3(url_destination, default_args['us_conn_id'], 'af-south-1', suffix)
+    for bucket, key, *rest in s3_inventory.list_keys():
+        destination_keys.append(key)
+
+    s3_inventory = s3(url_source, default_args['us_conn_id'], 'us-west-2', suffix)
+    for bucket, key, *rest in s3_inventory.list_keys():
+        if key.startswith('sentinel-s2-l2a-cogs') and len(key.split("/")) > 2 and \
+           key.split("/")[2].split("_")[1] in africa_tile_ids:
+            source_keys.append(key)
+
+    source_keys = set(source_keys)
+    destination_keys = set(destination_keys)
+    diff =  [[x] for x in (source_keys - destination_keys)]
+
+    output_filename = datetime.today().strftime("%d/%m/%Y %H:%M:%S") + ".json"
+    output_filename = output_filename.replace("/", "_").replace(":", "_").replace(" ", "_")
```

Comment:
Can we use `datetime.today().strftime("%d_%m_%Y_%H_%M_%S")` to avoid replace?

Reply:
This is better, thanks.